winpr/libwinpr/sspi/Negotiate/negotiate.c

```diff
@@ -625,83 +1331,92 @@ static SECURITY_STATUS SEC_ENTRY negotiate_QueryCredentialsAttributesA(PCredHand
 
 static SECURITY_STATUS SEC_ENTRY negotiate_FreeCredentialsHandle(PCredHandle phCredential)
 {
-	SSPI_CREDENTIALS* credentials;
+	MechCred* creds;
 
-	if (!phCredential)
+	creds = sspi_SecureHandleGetLowerPointer(phCredential);
+	if (!creds)
 		return SEC_E_INVALID_HANDLE;
 
-	credentials = (SSPI_CREDENTIALS*)sspi_SecureHandleGetLowerPointer(phCredential);
-
-	if (!credentials)
-		return SEC_E_INVALID_HANDLE;
+	for (int i = 0; i < MECH_COUNT; i++)
+		creds[i].mech->pkg->table->FreeCredentialsHandle(&creds[i].cred);
 
-	sspi_CredentialsFree(credentials);
-	sspi_SecureHandleInvalidate(phCredential);
 	return SEC_E_OK;
 }
 
 static SECURITY_STATUS SEC_ENTRY negotiate_EncryptMessage(PCtxtHandle phContext, ULONG fQOP,
                                                           PSecBufferDesc pMessage,
                                                           ULONG MessageSeqNo)
 {
-	NEGOTIATE_CONTEXT* context;
-	SECURITY_STATUS status = SEC_E_UNSUPPORTED_FUNCTION;
-	context = (NEGOTIATE_CONTEXT*)sspi_SecureHandleGetLowerPointer(phContext);
+	NEGOTIATE_CONTEXT* context = (NEGOTIATE_CONTEXT*)sspi_SecureHandleGetLowerPointer(phContext);
 
-	if (context->sspiW->EncryptMessage)
-		status =
-		    context->sspiW->EncryptMessage(&(context->SubContext), fQOP, pMessage, MessageSeqNo);
+	if (!context)
+		return SEC_E_INVALID_HANDLE;
+
+	if (context->mic)
+		MessageSeqNo++;
```

Comment:
Hum that's very strange that `MessageSeqNo` is passed by value

Reply:
Why's that?

Reply:
Ok I think I understand the thing, it's that when we have a MIC the real sequence number is always shifted by 1